src/screens/FilterModal/component.js

```diff
@@ -0,0 +1,153 @@
+// @flow
+import React, { PureComponent, Fragment } from "react";
+import { View, StyleSheet } from "react-native";
+import SafeAreaView from "react-native-safe-area-view";
+import type { NavigationScreenProp } from "react-navigation";
+import ShadowedScrollView from "../../components/ShadowedScrollView";
+import Button from "../../components/ButtonPrimary";
+import ContentPadding, {
+  getContentPadding
+} from "../../components/ContentPadding";
+import SectionHeader from "../../components/SectionHeader";
+import CheckBox from "../../components/CheckBox";
+import ScreenSizeProvider from "../../components/ScreenSizeProvider";
+import { bgColor } from "../../constants/colors";
+import tags from "../../data/tags";
+import text from "../../constants/text";
+import type { FilterCollection, Area } from "../../data/event-filters";
+import Header from "./Header";
+
+export type TagFilter = { [string]: Set<string> };
+
+type Props = {
+  navigation: NavigationScreenProp<{}>,
+  applyButtonText: string,
+  eventFilters: FilterCollection,
+  numEventsSelected: number,
+  numTagFiltersSelected: number,
+  onChange: TagFilter => void,
+  onApply: () => void,
+  onCancel: () => void
+};
+
+class FilterModal extends PureComponent<Props> {
+  componentDidMount() {
+    const { navigation, onCancel } = this.props;
+    this.didBlurSubscription = navigation.addListener("willBlur", onCancel);
+  }
+
+  componentWillUnmount() {
+    this.didBlurSubscription.remove();
+  }
+
+  didBlurSubscription: {
+    remove: () => void
+  };
+
+  toggleTagFilter = (
+    selectedValues: Set<string> | Set<Area>,
+    sectionValue: string
+  ) => {
+    const values: Set<string> = new Set([...selectedValues]);
+    if (!values.delete(sectionValue)) {
+      values.add(sectionValue);
+    }
+    return values;
+  };
+
+  clearTagFilters = () =>
+    this.props.onChange(
+      Object.keys(tags).reduce((acc, key) => ({ ...acc, [key]: new Set() }), {})
+    );
+
+  render() {
+    const {
+      applyButtonText,
+      onApply,
+      onChange,
+      eventFilters,
+      navigation,
+      numEventsSelected,
+      numTagFiltersSelected
+    } = this.props;
+    return (
+      <SafeAreaView
+        style={styles.flex}
+        forceInset={{ bottom: "always", top: "never" }}
+      >
+        <Header
+          onClearPress={this.clearTagFilters}
+          onCancelPress={() => navigation.goBack()}
+          showClear={numTagFiltersSelected > 0}
+        />
+        <ShadowedScrollView style={styles.flex} shadowOpacity={0.4}>
+          {Object.keys(tags).map(
+            sectionName =>
```

Comment:
The main reason for not using a SectionList was because Jun didn't want the sticky headers on this page. But I've only just noticed that `stickySectionHeadersEnabled` is a prop available on `SectionList`, so can use that.

Reply:
Oh, interesting, I thought those would be a nice UX bonus :) oh well

Reply:
...maybe the visual design should then be closer to this kind of thing: 

![](https://jira.appcelerator.org/secure/attachment/41863/Screen%20Shot%202013-08-26%20at%2009.44.43.png). 

(I think the only real difference is the alignment of text in the gap)

Reply:
Ok, I've broken up this into a couple of smaller components now, `FilterSectionList`, and `FilterList`. Looks cleaner and made the tests nicer too.

I looked into turning this into a `SectionList`, but I think it's going to turn quite messy and complicated putting it inside `ShadowedScrollView` as we'll have nested scroll views. I think it would require quite a big refactor of `ShadowedScrollView` to allow us to use its animated shadows around other elements. Luckily we don't really need the performance benefit here of the `SectionList` because it's such a short list. So I think it might be best to leave it as-is for now until the UX changes again.